### PR TITLE
os/bluestore/KernelDevice: helpful warning when aio limit exhausted

### PR DESCRIPTION
**Pull request by liewegas**
Suggested-by: Dan Van Der Ster <daniel.vanderster@cern.ch>
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
@dvanders

File: src/os/bluestore/KernelDevice.cc [hunk 1]
```diff
@@ -152,7 +152,9 @@ int KernelDevice::open(const string& p)
   size &= ~(block_size - 1);
 
   r = _aio_start();
-  assert(r == 0);
+  if (r < 0) {
```
Comment (ifed01) [inline src/os/bluestore/KernelDevice.cc]: as close() wouldn't be called after this failure one should delete/reset fs as well.

**Comment by dvanders** [comment]
@liewegas I'm confused. Per man io_setup we should check errno, not the return code.

**Comment by liewegas** [comment]
good catch.  reordered to avoid need for fs cleanup.

**Comment by liewegas** [comment]
<pre>
       Note that the libaio wrapper function uses a different type (io_context_t *) for the ctx_idp argument.  Note also that the libaio wrapper does not follow the usual C library  conventions  for  indicating
       errors: on error it returns a negated error number (the negative of one of the values listed in ERRORS).  If the system call is invoked via syscall(2), then the return value follows the usual conventions
       for indicating an error: -1, with errno set to a (positive) value that indicates the error.
</pre>
@dvanders

File: src/os/bluestore/KernelDevice.cc [hunk 2]
```diff
@@ -145,15 +145,17 @@ int KernelDevice::open(const string& p)
     }
   }
 
+  r = _aio_start();
+  if (r < 0) {
+    goto out_fail;
+  }
+
   fs = FS::create_by_fd(fd_direct);
```
Comment (ifed01) [inline src/os/bluestore/KernelDevice.cc]: Wondering if we need that at all, don't see any use of it

**Pull request merged**